src/mpyl/projects/versioning.py

```diff
@@ -0,0 +1,164 @@
+"""Versioning and upgrade utilities for mpyl projects.
+### Writing an upgrade script
+
+To write an upgrade script, create a new class that inherits from `Upgrader` and
+implements the `upgrade` method. This class should then be added to the `UPGRADERS`
+list in this module.
+
+"""
+import pkgutil
+from io import StringIO
+from pathlib import Path
+from typing import Optional, Generator
+import copy
+
+from deepdiff import DeepDiff
+from ruamel.yaml import YAML
+from ruamel.yaml.compat import ordereddict
+
+VERSION_FIELD = "mpylVersion"
+
+
+def get_releases() -> list[str]:
+    embedded_releases = pkgutil.get_data(__name__, "releases/releases.txt")
+    if not embedded_releases:
+        raise ValueError("File releases/releases.txt not found in package")
+    releases = embedded_releases.decode("utf-8").strip().splitlines()
+    return list(reversed(releases))
+
+
+def get_latest_release() -> str:
+    return get_releases()[0]
+
+
+def yaml_to_string(serializable: object, yaml: YAML) -> str:
+    with StringIO() as stream:
+        yaml.dump(serializable, stream)
+        return stream.getvalue()
+
+
+class Upgrader:
+    """Base class for upgrade scripts"""
+
+    target_version: str
+
+    def upgrade(self, previous_dict: ordereddict) -> ordereddict:
```

Comment:
Aren't dicts already ordered in python 3.9?

Reply:
This is a ruaml type. It has API for inserting fields.